c/compare.py

```diff
@@ -0,0 +1,174 @@
+(options, args) = parser.parse_args()
+
+# if the user did not directly specify sets, we select all sets
+SETS=[]
+for arg in (args if args else ["*.set"]):
+    SETS.extend(expand(arg))
+
+# build goto-cc and goto-diff if not available, and then set PATH to find it
+if not shutil.which("goto-cc") or not shutil.which("goto-diff"):
+  if not os.path.exists("../cbmc.git/src/goto-cc/goto-cc"):
+    subprocess.call(["git", "clone" "--depth=1" "http://github.com/diffblue/cbmc.git" "../cbmc.git"])
+    subprocess.call(["make", "minisat2-download"], cwd="../cbmc.git/src")
+    subprocess.call(["make", "CXX=g++-5", "goto-diff.dir", "goto-cc.dir"], cwd="../cbmc.git/src")
+  cwd = os.getcwd()
+  os.environ['PATH'] = cwd + "/../cbmc.git/src/goto-cc:" + cwd + "/../cbmc.git/src/goto-diff:" + os.environ['PATH']
+
+EC=0
+
+# iterate over all sets
+for f in SETS:
+  if not os.path.exists(f):
+    print("Invalid set", f)
+    exit(1)
+
+  setf=os.path.basename(f)[:-4] # remove ending ".set"
+
+  # skip some sets, like LDV (too big) or Concurrency (pthread headers are very platform dependent)
+  if setf == "ConcurrencySafety-Main":
+    print("Skipping category", setf, "(platform-dependent types)")
+    continue
+  elif options.SKIP_LARGE and setf.startswith("Systems_DeviceDriversLinux64_ReachSafety"):
+    print("Skipping category", setf, "(only custom includes, no system headers, checking takes too much time)")
+    continue
+  elif setf == "Systems_OpenBSD_MemSafety":
+    print("Skipping category", setf, "(only custom includes, no system headers, complicated build process)")
+    continue
+  elif setf == "Systems_SQLite_MemSafety":
+    print("Skipping category", setf, "(complicated build process, requires patched version of cilly)")
+    continue
+
+  if not os.path.exists(setf + ".cfg"):
+    print("Skipping category", setf, "(no .cfg file present)")
+    continue
+
+  print("Processing category", setf)
+  bits=getArchitecture(setf + ".cfg")
+  if bits not in ["32", "64"]:
+    print("Invalid bit width in file", setf + ".cfg")
+    exit(1)
+
+  # iterate over all files in the set
+  i=0
+  for ff in getTasksFromSet(f):
+    orig=ff
+
+    if ff.endswith(".yml"):
+      with open(ff, 'r') as yamlfile:
+        yml = yaml.safe_load(yamlfile)
+
+      # check whether the input_basename exists (nobody should ever use "null" as a filename!)
+      inputFiles = yml['input_files']
+      if not inputFiles:
+        print("No input files defined in", ff)
+        exit(1)
+
+      # check whether there is exactly one input file, either directly or nested in a list
+      if isinstance(inputFiles, list):
+        print("ignoring task consisting of multiple sourcefiles")
+        continue
+      ff = os.path.join(os.path.dirname(ff), inputFiles)
+
+    if not os.path.exists(ff):
+      print("No task file", ff, "found")
+      exit(1)
+
+    # no original source available
+    if ff.startswith(('ddv-machzwd/', 'aws-c-common/', 'ldv-linux-3.0/', 'ldv-regression/')):
```

Comment:
It is somewhat inconsistent that some files are blacklisted, and some files are not even checked. Could this be merged? If we need both, then there should be two lists of patterns at the start of the script, not one hidden in the middle of the script.